### ISSUE TITLE
implement expensive Schur complement solve

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#921]
I have some working code, but it needs cleaning up.

@egpuckett You had somebody working on solvers or a very difficult test, right?

**Comment by bangerth** [comment]
Specifically, what @tjhei says he tried is not to approximate `S` by `M_p` (the mass matrix on the pressure space, scaled by the inverse of the viscosity), but to approximate `S` by `(B^T V(A) B)` where `V(A)` corresponds to one V-cycle on the matrix A, thereby approximating `A^{-1}`.

In the preconditioner, one needs to apply `S^{-1}`. We currently approximate this by running a CG on `M_p`. In @tjhei 's scheme, one would run a CG method on `(B^T V(A) B)`, and precondition it with `ILU(M_p)` (which we currently already have).

It should all be relatively straightforward to implement.

**Comment by yinghe616** [comment]
@egpuckett , @magali, and I were planning to implement an expensive Schur complement.  Yes, we wanted to approximate S by  (B^T V(A) B), but precondition it with some approximation of (B^T V(A)^{-1} B), for example  (B^T diag(V(A)^{-1}) B)  not just  ILU(M_p).  @tjhei , did you already implement other preconditioner for S?

**Comment by bangerth** [comment]
I'm almost certain that `ILU(M)` is good enough, and in practice much easier to implement. It should be good enough because `M` approximates `S=B^T A B`, and therefore also approximates `B^T V(A) B`.

**Comment by yinghe616** [comment]
I haven't implemented and tested yet, but in Dave A. May and Louis Moresi 's 2008 paper, they have studied a lot on how the total CG iteration number can be reduced due to different choice on the approximation of (B^T V(A)^{-1} B)...  I agree with you that ILU(M) is much easier to implement . But in case ILU(M) is not good enough for the large jump viscosity case,  it's still worth to try the other preconditioners?

**Comment by bangerth** [comment]
Sure, it's worth it. But it's worth trying the simplest approach first and see whether that actually works :-)

**Comment by yinghe616** [comment]
Yes, of course we should always start from the simplest approach first. And good to know that @tjhei already implemented it,  I think we don't need to try it anymore. :)

**Comment by tjhei** [comment]
> It should all be relatively straightforward to implement.

yes, already exists here (mingled together with the null space stuff):
https://github.com/tjhei/aspect/commit/e5c7bd9a05ba75d104260ab5969e4a1a28e64e24
